### DISCUSSION TITLE
Networking Requirements for CNFs

### DISCUSSION THREAD

**Discussion by vukg** [lfn-cnti/bestpractices#37] [Ideas]
I would like to invite here my colleague @szengerle to provide us some inputs about CNF networking.

**Comment by taylor** [comment]
Welcome, @szengerle. Also please join #cnf-wg on http://slack.cncf.io/

**Comment by vukg** [comment]
Actually, until @szengerle gets active I wanted to copy some requirements that he is driving in our internal activities. Please see below:

Networking
----------

In recent years, 3GPP has been driving Control and User Plane Separation (CUPS) for VNFs. The CUPS architecture splits monolithic network functions into their control and user plane tasks. The control plane components terminate control plane protocols such as SIP (TCP/UDP), Diameter (SCTP), HTTP(S) (TCP/UDP) and GTP-C (UDP). The actual forwarding, routing, and other network functions are performed in the user plane components. These must support protocols on different OSI layers (e.g., MAC/Layer 2, IP/Layer 3, RTP/Layer 7 and GTP-U/Layer 7). Since these sometimes perform routing, forwarding or even transcoding between different interfaces and network sections, they are usually divided among several network interfaces.

CNFs with a CUPS architecture can be divided into pods that have only control plane functionality and those with user plane functionality. Control plane functions are no different from pure IT applications, except that they often use typical telco protocols. Such are e.g., Diameter over SCTP or SIP. Both protocols require additional NATing efforts. However, no exception should be made for control plane components regarding typical Kubernetes reuse patterns. Telco control plane functions must also support pure primary networking without multiple interfaces. If numerous interfaces are required, segregation of the function into microservices is recommended.
Furthermore, services and ingress must be used to expose the functions. The usage of services and ingress ensures that CNF use the services provided by the platform, e.g. load balancer, IPAM, service discovery. Using Kubernetes orchestration for IPAM and service discovery avoids the historical use of complex and error-prone IP planning processes.

- **NW-1 CUPS**: The vendor shall make use of the CUPS architecture to divide pods with control- and such with user-plane functionality
- **NW-2 Primary Networking**: The vendor shall only use primary networking for control-plane functions.
    - **NW-2-A**: The vendor shall use Ingress resources to expose HTTP based control-plane interfaces.
    - **NW-2-B**: The vendor shall use Services to expose non HTTP control-plane interfaces.
    - **NW-2-C**: The vendor shall supported NATed pod ports to external IP addresses (Service make use of NAT to expose Pod Ports. Some protocols such as Diameter specify there external IP address in headers such as the Diameter AVP Host-IP-Address.)

For user plane functions, the use of secondary interfaces via Multus is permitted in exceptional cases. For example, if the function performs forwarding/routing on Layer 3, multiple subnets can be connected to the function. Layer 2 network functions are another exception. SRIOV via secondary interface is furthermore supported.  The network attachment definition for secondary networks is defined with the vendor during onboarding. Network functions should delegate IPAM functionality to Kubernetes for Layer 3 interfaces by requesting IPAM via network attachment definition. This involves assigning unique IP addresses to pods across the cluster. The control plane functions must ensure service discovery based on the addresses defined by Kubernetes. Using Kubernetes-based IPAM eliminates manual IP planning processes. In particular, the announcement of user-plane functions to third-party platforms and systems must be ensured by the control plane. This concerns, for example, the transmission of the correct GTP-U address of a SGW to the eNB/gNB during the initial attachment via S1AP or the RTP address of the MGW to CPEs via SIP/SDP.

- **NW-3** Secondary Networking: The vendor can use Multus secondary networking for user-plane functions.
    - **NW-3-A**: The vendor can use Multus secondary networking for SRIOV user-plane.
    - **NW-3-B**: The vendor shall make use of Kubernetes IPAM functionality for user-plane secondary interfaces
    - **NW-3-C**: The vendor shall implement internal service discovery for user-plane secondary interfaces inside their control-plane.
    - **NW-3-D**: The vendor shall make use of their control-plane to announce dynamically IP addresses of secondary interfaces to peers accessing the functions’ control plane

I am interested to hear some opinions from CNF dev side @CsatariGergely @iawells @rannyh @electrocucaracha and co.

**Reply by rannyh** [reply]
Thanks @vukg , this is a good starting point. I am a bit concerned about the explicit reference of Multus. There are other alternative to CNI multiplexing, and the cloud infrastructure may or may not have them available. There was another discussion thread where the idea of decoupling the CNF networking requirements from the platform capabilities was raised. The CNF needs to have a mechanism to indicate what are its preferred network requirements, and the infrastructure should have a mechanism to indicate the supported network capabilities, so that an orchestrator can optimally place the CNF. 
Anyway, bottom line is that networking requirements of CNF should be more abstract, and avoid requiring a specific CNI plugin.

**Reply by vukg** [reply]
Sure, that was direct copy paste from our setup which is in parts opinionated in order to be able to reasonably run a managed service. But it should be more generic in general.

**Reply by electrocucaracha** [reply]
Hola @vukg , thanks for adding me here as well. 

I have the impression that portability is violated when operators have to pre-define secondary networks during on-boarding and leave IP assignment to K8s. I have seen that prerequisite in Multus and DANM but not in NSM so as @rannyh mentioned maybe removing Multus from the equation can facilitate some operation tasks. 

> For user plane functions, the use of secondary interfaces via Multus is permitted in exceptional cases. For example, if the function performs forwarding/routing on Layer 3, multiple subnets can be connected to the function. Layer 2 network functions are another exception. SRIOV via secondary interface is furthermore supported. **The network attachment definition for secondary networks is defined with the vendor during onboarding. Network functions should delegate IPAM functionality to Kubernetes for Layer 3 interfaces by requesting IPAM via network attachment definition.** This involves assigning unique IP addresses to pods across the cluster. The control plane functions must ensure service discovery based on the addresses defined by Kubernetes. Using Kubernetes-based IPAM eliminates manual IP planning processes. In particular, the announcement of user-plane functions to third-party platforms and systems must be ensured by the control plane. This concerns, for example, the transmission of the correct GTP-U address of a SGW to the eNB/gNB during the initial attachment via S1AP or the RTP address of the MGW to CPEs via SIP/SDP.

**Comment by CsatariGergely** [comment]
In the NW-2-B case a loadbalancer will require multiple interfaces also in case of SCTP MH multiple interfaces are required.

**Reply by szengerle** [reply]
Thanks for the comment. 

My idea was to use a service (load balancer) with an attached externalIP per SCTP endpoint. Having multiple SCTP interfaces on pods brings no additional redundancy, in my opinion. The data center fabric and host (running pod) are still dedicated. No additional redundancy can be provided by having multiple interfaces on pods. Multiple services can provide multiple endpoints to external components such as eNB/gNB. This redundancy makes sense e.g., to have a disjunct path in the backbones.

What do you think about the proposal?


```
            +------------+
ext.+-------+ Service A  +----+  Single IP / Interface
IP A        +------------+    |     +-----------+
                              +-----+  Pod      |
            +------------+    |     +-----------+
ext.+-------+ Service B  +----+
IP B        +------------+

```

**Reply by manuelstein** [reply]
Just wanted to make a note of the discussion in https://github.com/intel/multus-cni/issues/466. 

My personal opinion on this is: the differentiation between primary and additional interfaces is a good, pragmatic approach to build on top of things, but consider the case when a network attachment is owning the endpoint that I want my SCTP app to be homed at. I would find that redirecting the traffic through an externalIP load balancer to the primary interface a bit of a detour.
So, if the `IP B` belongs in the IPAM of a NetworkAttachment (that I also want the pod to be homed in through a secondary IF), should I have to use a Service with an external LB for the SCTP traffic so that gets NATed to my pod's primary IF even though I have a secondary interface in the same subnet?

There's probably enough use cases to justify both setups (expose a network function through service resources or multiple interfaces) and the solution should be flexible enough.

Regarding the UC, I am used to having a _single_ service expose a _set_ of pods, and I haven't seen it the other way around. Could the case you're showing also be designed with a single service definition backed by a controller that has LBs in multiple networks? IIRC, MH SCTP requires port numbers to match across IPs and having two resource objects that need to be in sync seems like it could complicate things. A K8s Service's status field for ingress IP is a list and could contain more than one IP, so why use two Services?

**Reply by manuelstein** [reply]
found the fitting question asked by @Levovar in [the DANM readme](https://github.com/nokia/danm#our-philosophy-and-motivation-behind-danm):
> Why couldn't NetworkPolicies, Services, LoadBalancers, all of these existing and proven Kubernetes constructs work with all network interfaces?

**Reply by Levovar** [reply]
that's not how -real- SCTP MH works
the endpoint itself needs to own multiple physical interfaces, as the SCTP associations are actively being established based on this knowledge:
"Each SCTP end point needs to check reachability of the primary and redundant addresses of the remote end point using a heartbeat. Each SCTP end point needs to acknowledge the heartbeats it receives from the remote end point."

Sure you can add an SCTP loadbalancer in front of your Pods which terminates both primary and sedonary path, but guess what: now this LB Pod needs to have 2 IPs

**Reply by manuelstein** [reply]
I had the same thought about the LB, it's the next step. Having a multi-homed, externalLB is shifting the problem outside K8s, but I'd want to see the LB run as a CNF itself.

**Comment by szengerle** [comment]
Thanks @vukg for adding me to this discussion.
